### PR TITLE
Implement support for IPv6 GW addresses in management domains

### LINKED ISSUE #232: ipv6_gw in to managament domain
**Is your feature request related to a problem? Please describe.**
We need the ability to run IPv6 in the management domain. Management domain does not allow to set ipv6_gw as a field and then we are unable to generate a management vxlan with IPv6-adresses. 

**Describe the solution you'd like**
It should be possible to add a ipv6_gw field to the management domain. 

**Describe alternatives you've considered**
Deriving ipv6_gw from the ipv4-address (ipv4_gw) is not at good solution for us.

### PR DESCRIPTION
**Pull request by lunkwill42**
Fixes #232 according to the design proposal of that issue.

However, I do not consider this entirely complete yet.  Though I have added some automated tests, I have not put my changes to the test in a lab thus far.  

I have a few comments and a few questions for followup, which I will post separately below.

### CODE REVIEW

File: src/cnaas_nms/api/mgmtdomain.py [hunk 1]
```diff
@@ -53,6 +55,23 @@ def ipv4_gw_valid_address(cls, v, values, **kwargs):
...
+            addr = IPv6Interface(v)
+            prefix_len = int(addr.network.prefixlen)
+        except:
+            raise ValueError('Invalid ipv6_gw received. Must be correct IPv6 address with mask')
+        else:
+            if addr.ip == addr.network.network_address:
+                raise ValueError("Specify gateway address, not subnet address")
+            if addr.ip == addr.network.broadcast_address:
+                raise ValueError("Specify gateway address, not broadcast address")
+            if prefix_len >= 126 or prefix_len <= 63:
```
Comment (lunkwill42) [inline src/cnaas_nms/api/mgmtdomain.py:70]: These values are my wild assumptions. Some feedback from @vidarstokke would be appreciated :)

Comment (vidarstokke) [reply src/cnaas_nms/api/mgmtdomain.py:70]: I agree. These are good values.

**Comment by lunkwill42** [comment]
1. I find the implementation of `init_access_device_step1` to be overly complex: It was 208 lines before I made my additions, and it became 262 lines afterwards. Both numbers are an extreme code smell, IMNSHO. 
  
   Shouldn't we be looking for ways to break this down into more manageable parts? (Something I would deem a separate cleanup PR, this PR is already becoming big).

2. Some of the variable stuff that `init_access_device_step1` works on seems to overlap with this implementation: https://github.com/SUNET/cnaas-nms/blob/5dcb14e2d102b42170f9e0da865bfe64a821e1ae/src/cnaas_nms/devicehandler/sync_devices.py#L137

   Why the redundancy? I'm thinking the latter also needs to change if I want secondary addresses to always be available to templates?

3. I'm not sure I understand what this part is used for: https://github.com/SUNET/cnaas-nms/blob/151c150cd969a381abb6d93e7c13dba5875a59d0/src/cnaas_nms/devicehandler/init_device.py#L571-L578
   
   My initial thought is that this plugin system needs to change to also be able to be notified about IPv6 changes?

File: src/cnaas_nms/api/mgmtdomain.py [hunk 2]
```diff
@@ -29,6 +29,7 @@
         "device_b": fields.String(required=True),
         "vlan": fields.Integer(required=True),
         "ipv4_gw": fields.String(required=True),
+        "ipv6_gw": fields.String(required=True),
```
[the remark this answers is not in the document]
Comment (lunkwill42) [reply src/cnaas_nms/api/mgmtdomain.py:32]: Don't ask me how Flask-RESTX works, I'm a complete n00b there. 

I'm not sure what these `required` parameters really affect. I'm still able to post management domains payloads with just one of these values present - it seems the only code that checks for their presence is this bit:

https://github.com/SUNET/cnaas-nms/blob/eb30a2ff89bce325ff56097e4b56a5e2e1337cd4/src/cnaas_nms/api/mgmtdomain.py#L186-L220

File: src/cnaas_nms/api/mgmtdomain.py [hunk 3]
```diff
@@ -56,6 +58,24 @@ def ipv4_gw_valid_address(cls, v, values, **kwargs):
...
+            addr = IPv6Interface(v)
+            prefix_len = int(addr.network.prefixlen)
+        except Exception:  # noqa: S110
+            raise ValueError("Invalid ipv6_gw received. Must be correct IPv6 address with mask")
+        else:
+            if addr.ip == addr.network.network_address:
+                raise ValueError("Specify gateway address, not subnet address")
+            if addr.ip == addr.network.broadcast_address:
+                raise ValueError("Specify gateway address, not broadcast address")
+            if prefix_len >= 126 or prefix_len <= 63:
```
[the remark this answers is not in the document]
Comment (lunkwill42) [reply src/cnaas_nms/api/mgmtdomain.py:74]: Indeed, but I opted to use the same test patterne as the pre-existing `ipv4_gw_valid_address`, for the sake of a little consistenct.

File: src/cnaas_nms/api/mgmtdomain.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lunkwill42) [reply src/cnaas_nms/api/mgmtdomain.py:32]: Perhaps this model is there just for the auto-generated Swagger docs? It's seems maybe the `expect` decorator in the following instance would need a `validate` argument if it were to automatically validate the input according to the model?

https://github.com/SUNET/cnaas-nms/blob/eb30a2ff89bce325ff56097e4b56a5e2e1337cd4/src/cnaas_nms/api/mgmtdomain.py#L153-L154

How would you specify a model where at least one of the two attributes are required to be present?

**Comment by indy-independence** [comment]
> 1. I find the implementation of `init_access_device_step1` to be overly complex: It was 208 lines before I made my additions, and it became 262 lines afterwards. Both numbers are an extreme code smell, IMNSHO.
>        Shouldn't we be looking for ways to break this down into more manageable parts? (Something I would deem a separate cleanup PR, this PR is already becoming big).

Yes I agree it's too long, I think the complexity kind of exploded when we added support for MLAG-pairs and I suspect it will get a lot worse when trying to support switch stacks also. There's also so many input arguments to the function, and I imagine breaking out parts of the code might require passing a long list of arguments as well. Another reason it's harder to refactor this code is the testing, where in our virtual labs we don't have all these variants with MLAG pair, MLAG pair with second level of access switch underneath, single connected access switch etc so making changes requires manual testing in physical lab. If we had one FTE just setting up the matrix of all types of device_inits with all types of platforms maybe we could get somewhere :)

 
>     2. Some of the variable stuff that `init_access_device_step1` works on seems to overlap with this implementation: https://github.com/SUNET/cnaas-nms/blob/5dcb14e2d102b42170f9e0da865bfe64a821e1ae/src/cnaas_nms/devicehandler/sync_devices.py#L137
>        
>        Why the redundancy? I'm thinking the latter also needs to change if I want secondary addresses to always be available to templates?

Hm I think there was an attempt to break this out into a function called populate_devicevars or something similar, but maybe not everything made it in there

>     3. I'm not sure I understand what this part is used for: https://github.com/SUNET/cnaas-nms/blob/151c150cd969a381abb6d93e7c13dba5875a59d0/src/cnaas_nms/devicehandler/init_device.py#L571-L578
>        
>        My initial thought is that this plugin system needs to change to also be able to be notified about IPv6 changes?

Yes I think the plugin api thing needs to be modified to also include ipv6 or secondary_address

>     4. I still don't understand why `nms.sql` and `nms_test.sql` are needed, and I don't understand how to update them when I make changed to the database schema (I've done so by manually editing them in a separate commit here). Shouldn't Alembic create the initial schema, and then test fixtures (as I assume `nms_test.sql` represents) should contain only data, not schema?

I think maybe the postgres needs an inital sql file to start up maybe, and then alembic updates it to the latest schema? I might need to ask kristofer why it looks like this

**Comment by lunkwill42** [comment]
> Yes I agree it's too long, I think the complexity kind of exploded when we added support for MLAG-pairs and I suspect it will get a lot worse when trying to support switch stacks also. 

I guess we need to start with baby steps then :)

> Hm I think there was an attempt to break this out into a function called populate_devicevars or something similar, but maybe not everything made it in there

That's exactly what I'm pointing at. There is `populate_device_vars()`, and the init step sort of first populates somee vars of its own, and then delegates the rest to `populate_device_vars()`, which has some overlap with what the init step did. Sounds to me like something could/should be removed from the init step, as `populate_device_vars()` is meant to be re-usable.

> Yes I think the plugin api thing needs to be modified to also include ipv6 or secondary_address

The mentioned hook is explicitly called `allocated_ipv4`. I guess I could add an `allocated_ipv6` - but is there a reason why this needs to be ip-version specific at all?

> I think maybe the postgres needs an inital sql file to start up maybe, and then alembic updates it to the latest schema? I might need to ask kristofer why it looks like this

The PostgreSQL Docker image doesn't require an initial SQL file, it is optional. What you may really want is for the container to ensure a specific database/user/password combo is initialized, so the database is ready to be filled with schema and data.

I would suppose that alembic is entirely capable of initializing the schema from scratch. I would find that preferable to today's solution - the current solution builds an entirely new PostgreSQL image from one that is initialized with `nms.sql`, which complicates our build pipeline, since we now have to publish our own dedicated PostgreSQL image for every new version of CNaaS-NMS.

Anyways, I've just proposed #288 as a mitigation that solves the current test issues. Merging this would allow me to strip all my `.sql` file changes from this PR. It also allows us to keep `nms.sql`&co as-is, until we can perhaps agree on a better way...

**Comment by lunkwill42** [comment]
Well alright, thanks for merging #288 so quickly :) I've stripped the sql fixture changes from this PR and rebased it on the latest `develop` HEAD.

File: src/cnaas_nms/api/mgmtdomain.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (indy-independence) [reply src/cnaas_nms/api/mgmtdomain.py:32]: Yes exactly I think it's for swagger docs only

**Comment by indy-independence** [comment]
So this should be fine to merge now? And we have some stuff to handle in future PRs but it's kind of separate from this?

**Comment by lunkwill42** [comment]
> So this should be fine to merge now? And we have some stuff to handle in future PRs but it's kind of separate from this?

It works from my point of view, but I have yet to receive proper feedback from our network engineers. I would say that this is ready to merge, as long as it doesn't break any existing functionality for you.

If feedback from our team suggests we need more changes, we can make new PRs.

Also, some of the IPv6-related code additions made in this PR would be useful to have there for our other upcoming IPv6-related PRs for linknets and infra IPs.

**Comment by lunkwill42** [comment]
Actually, I forgot to remove some of the unnecessary SQL changes mentioned above, so I've rebased and force-pushed the entire branch to ensure they were properly removed.

**Comment by lunkwill42** [comment]
Rebased on develop and force-pushed again, by tip from @indy-independence that the integration test suite had been accidentally disabled in the develop branch before yesterday.

**Pull request merged**